docs/contributing/plugin-tutorial.md

```diff
@@ -4,268 +4,199 @@ Haskell Language Server is an LSP server for the Haskell programming language. I
+Using the convenience `defaultPluginDescriptor` function, the plugin can be bootstrapped with the required parts:
 
 ```haskell
-idePlugins = pluginDescToIdePlugins allPlugins
-  where
-    allPlugins =
-      [ GhcIde.descriptor "ghcide"
-      , Pragmas.descriptor "pragmas"
-      , Floskell.descriptor "floskell"
-      , Fourmolu.descriptor "fourmolu"
-      , Ormolu.descriptor "ormolu"
-      , StylishHaskell.descriptor "stylish-haskell"
-      , Retrie.descriptor "retrie"
-      , Eval.descriptor "eval"
-      ]
+-- plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs
+
+-- | The "main" function of a plugin
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId =
+  (defaultPluginDescriptor plId)
+    { pluginCommands = [importLensCommand], -- The plugin provides a command handler
+      pluginHandlers = mconcat -- The plugin provides request handlers
+        [ lensProvider
+        ]
+    }
 ```
-To add a new plugin, simply extend the list of `allPlugins` and rebuild.
-
-## Providers
-
-99% of plugins will want to define at least one type of provider. But what is a provider? Let's take a look at some types:
-```haskell
-type CodeActionProvider =  LSP.LspFuncs Config
-                        -> IdeState
-                        -> PluginId
-                        -> TextDocumentIdentifier
-                        -> Range
-                        -> CodeActionContext
-                        -> IO (Either ResponseError (List CAResult))
 
-type CompletionProvider = LSP.LspFuncs Config
-                        -> IdeState
-                        -> CompletionParams
-                        -> IO (Either ResponseError CompletionResponseResult)
+We'll start with the command, since it's the simplest of the two.
 
-type CodeLensProvider = LSP.LspFuncs Config
-                      -> IdeState
-                      -> PluginId
-                      -> CodeLensParams
-                      -> IO (Either ResponseError (List CodeLens))
+### The command handler
 
-type RenameProvider = LSP.LspFuncs Config
-                    -> IdeState
-                    -> RenameParams
-                    -> IO (Either ResponseError WorkspaceEdit)
-```
+In short, commands works like this:
+- The LSP server (HLS) initially sends a command descriptor to the client, in this case as part of a code lens.
+- Whenever the client decides to execute the command on behalf of a user action (in this case a click on the code lens), it sends this same descriptor back to the LSP server which then proceeds to handle and execute the command. The latter part is implemented by the `commandFunc` field of our `PluginCommand` value.
 
-Providers are functions that receive some inputs and produce an IO computation that returns either an error or some result.
+> **Note**: Check the [LSP spec](https://microsoft.github.io/language-server-protocol/specification) for a deeper understanding of how commands work.
```

Comment:
Another reason I like RST: proper support for admonitions like this.

Reply:
I created an [issue](https://github.com/haskell/haskell-language-server/issues/3663) to track this.